CARDS-2514: New "locking" module, with basic locking feature

acrowthe opened the pull request:
Initial implementation including:
- Locking restriction
- Lock node on subjects
  - Includes author who locked the node and time
  - Only locked subjects with a child lock node can be unlocked and only if their parent is unlocked
    to prevent partially unlocked child trees
- Basic UI for locking and unlocking
  - Single lock/unlock button based on current lock state

Testing:
- Verify that locking is opt-in:
  - Launch using `./start_cards.sh --test --dev` to make sure nothing has changed and that locking is disabled by default
  - Launch using `./start_cards.sh --test --dev --locking` to test that locking is can be enabled and is working as expected
  - Prepare and run docker with locking:
    - `python3 generate_compose_yaml.py --cards_project cards4prems --oak_filesystem --dev_docker_image --locking`
    - `docker-compose build && docker-compose up`
    - Test that locking is working as expected
    - `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
  - Prepare and run docker without locking: `python3 generate_compose_yaml.py --cards_project cards4prems --oak_filesystem --dev_docker_image`

Make sure that:
- Existing action buttons on the subject page work as before (delete, print preview)
- When a subject is locked, non-admin users cannot edit the subject or any of it's child forms
  - Currently any user can add or remove locks, this will be restricted by a future task
- Subjects cannot be locked or unlocked if they have a locked parent
- If you lock a child subject then lock a parent subject, unlocking the parent will not unlock the child
- If you lock a parent subject that has unlocked children (forms and/or subject), the child nodes will be locked too

https://github.com/data-team-uhn/cards/assets/20056751/b737e737-6a69-4e40-b732-7b0dda7e280b

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
Initial implementation including:
- Locking restriction
  - TODO: Once lock and unlock rights are implemented, apply locked restriction to everyone by default
- Lock node on subjects
  - Includes author who locked the node
  - Only locked subjects with a child lock node can be unlocked and only if their parent is unlocked
    to prevent partially unlocked child trees
  - TODO: add time to lock node
- Basic UI for locking and unlocking
  - Single lock/unlock button based on current lock state
  - TODO: list any incomplete forms
  - TODO: update UI when locking or unlocking successfull to reflect new lock state
  - TODO: list current user and timestamp

Other tasks TODO:
- Make locking opt-in
- Make subject actions extensible to move the locking button into the locking modules

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
Lock button will now update the subject view so changes to locked status will
be displayed to the user without needing to refresh the page

acrowthe commented:
Improved UI so it updates subject data whenever the locking dialog is completed

https://github.com/data-team-uhn/cards/assets/20056751/172bd0e8-1414-4295-ad07-3405dcb9774a

sdumitriu commented on distribution/src/main/features/core/locking.json:
This should be deleted, the feature is supposed to be optional.

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockServlet.java:
```diff
@@ -0,0 +1,287 @@
+@Component(service = { Servlet.class })
+@SlingServletResourceTypes(
+    resourceTypes = { "cards/Subject" },
+    methods = { "POST" })
```
Can you try to use the `LOCK` and `UNLOCK` methods instead of `POST`?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -429,23 +463,31 @@ function SubjectMemberInternal (props) {
-                   className={classes.childSubjectHeaderButton}
-                 />
-               </>
+                <LockButton
```
This makes it always enabled as a core feature. Good for the demo, but the list of actions should be turned into extensions.

marta- replied:
that's https://phenotips.atlassian.net/browse/CARDS-1084

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -429,23 +463,31 @@ function SubjectMemberInternal (props) {
+                  entryName={identifier}
+                  statusFlags={statusFlags}
+                  className={classes.childSubjectHeaderButton}
+                  onComplete={fetchSubjectData}
```
IIRC, in the form UI there is a way to fetch and update just the status flags. Does it have to fetch the entire subject, or can we also just fetch the status flags here?

marta- replied:
We should refetch only the status flags, but it looks like that requires a bunch of refactoring. This is fine for the demo, but should be done before merge,

acrowthe replied:
In this case status flags would work fine - when this is refactored to have actions be extensible I think giving actions the option to refresh the whole subject would be more useful in other cases

acrowthe replied:
Going to keep the resfresh action to the whole subject as it is not just the subject's statusFlags that are updated:
- Child forms/subjects statusFlags
- the current subject's lock node

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/LockButton.jsx:
```diff
@@ -0,0 +1,192 @@
+    if (isLocked) {
+      setDialogMessage("Are you sure you would like to unlock this " + entryType + "?");
+    } else {
+      setDialogMessage("Would you like to lock this " + entryType + "?");
```
We should rephrase this to mention "signoff". I'm aware that i didn't mention it in the specs, but now that I see the UI the absence of "signoff" stands out.

Something along the lines of:
* Title: "Sign off and lock"
* Dialog content:
```
Signed off by: <user name>
Date: <now>
<Alert severity="warning">Once you sign off on this <entrytype>, it will be locked along with all the associated data forms, and no further edits will be possible. Proceed?</Alert>
```

**This would be very good to have for the demo.**

acrowthe replied:
Done

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
Switch to LOCK and UNLOCK HTTP methods rather than Post

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
- Add username and date to lock dialog

acrowthe commented:
Updated lock and unlock dialogs:
![image](https://github.com/data-team-uhn/cards/assets/20056751/97087eb4-94bb-4d1b-b60b-c542b3bc71ff)
![image](https://github.com/data-team-uhn/cards/assets/20056751/d7bf15fc-7aca-4a83-8da4-77d1b9afe676)

Changed HTTP method from POST to LOCK and UNLOCK

sdumitriu commented on modules/locking/src/main/features/feature.json:
```diff
@@ -0,0 +1,41 @@
+        // "set ACL for everyone \n deny rep:write,jcr:versionManagement on /Forms,/Subjects restriction(cards:locked) \n end"
+      ]
+    },
+    "org.apache.sling.serviceusermapping.impl.ServiceUserMapperImpl.amended~patientUi":{
```
```suggestion
    "org.apache.sling.serviceusermapping.impl.ServiceUserMapperImpl.amended~locking":{
```

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
Create new Subject Action extension
- Move lock button into locking module and access it through the extension

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
Fix incorrectly named configuration

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
- Make locking opt-in

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
- Restrict form and child subject creation for locked subjects

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
- Improve locking UI
  - Simplify subject actions extension
  - Add list of incomplete forms to lock UI
  - Update child form display when a subject is locked or unlocked

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
Improve locking UI
- Split locking dialog into up to 2 dialogs:
  - List of incomplete forms + continue
  - Sign off
- Convert username and date into disabled inputs
- Convert list of incomplete forms into Material UI list

acrowthe commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockServlet.java:
```diff
@@ -0,0 +1,324 @@
+    }
+
+    // TODO: Clean up
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity"})
```
This needs to be improved before merging

acrowthe commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockServlet.java:
```diff
@@ -0,0 +1,324 @@
+            final VersionManager versionManager = session.getWorkspace().getVersionManager();
+
+            final Node requestNode = request.getResource().adaptTo(Node.class);
+            // TODO: check if user has permission to lock and unlock the request node
```
This should be implemented in a separate issue that adds rights to lock and unlock resources

marta- commented on modules/locking/src/main/frontend/src/locking/SubjectLockAction.jsx:
```diff
@@ -99,41 +99,101 @@ function SubjectLockAction(props) {
+        setDialogContent(
+          <>
+            <TextField
+              disabled
```
For both textfields, I would replace `disabled` with `inputProps={{readonly:true}}` so they look less faded

sashaandjic commented:
Suggestion: Add `LOCAKED BY ...`
![image](https://github.com/data-team-uhn/cards/assets/23243244/8e4ac408-2d18-4dd2-a688-b5ae6d74b332)

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -577,6 +577,10 @@ const questionnaireStyle = theme => ({
     },
+    LOCKEDChip:{
+      backgroundColor: theme.palette.primary.main,
+      color: "#FFF"
```
Could this be made more flexible? Maybe paper color, or the opposite of text color?

acrowthe replied:
Changed to background paper color

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/content/apps/cards/ExtensionPoints/SubjectActions.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "jcr:primaryType": "cards:ExtensionPoint",
+  "cards:extensionPointId": "cards/coreUI/subjectActions",
```
I'd name this `cards/subject/chart/actions` instead. And I wonder if it should be defined in `modules/data-model/subjects/api/`.

sdumitriu commented on modules/locking/src/main/resources/SLING-INF/nodetypes/locking.cnd:
```diff
@@ -0,0 +1,44 @@
+  - sling:resourceSuperType (STRING) = "cards/Resource" mandatory autocreated protected
+
+  // The time that the parent node was locked
+  - time (date) mandatory
```
```suggestion
  - time (DATE) mandatory
```

We should all try to be more consistent with this syntax...

sdumitriu commented on modules/locking/src/main/resources/SLING-INF/nodetypes/locking.cnd:
```diff
@@ -0,0 +1,44 @@
+  // Properties:
+  // Hardcode the resource type
+  - sling:resourceType (STRING) = "cards/Lock" mandatory autocreated protected
+  // Hardcode the resource supertype: the login page is a resource
```
Almost every time a new node is defined, this comment is copy-pasted without being updated with the right comment. Maybe it should be simplified so it doesn't need updating.

sdumitriu commented on modules/locking/src/main/resources/SLING-INF/nodetypes/locking.cnd:
```diff
@@ -0,0 +1,44 @@
+  // Hardcode the resource type
+  - sling:resourceType (STRING) = "cards/Lock" mandatory autocreated protected
+  // Hardcode the resource supertype: the login page is a resource
+  - sling:resourceSuperType (STRING) = "cards/Resource" mandatory autocreated protected
```
This isn't a resource. It should be a `cards:Item`, but [that PR](https://github.com/data-team-uhn/cards/pull/1648) is still a draft.

sdumitriu commented on modules/locking/src/main/resources/SLING-INF/nodetypes/locking.cnd:
```diff
@@ -0,0 +1,44 @@
+//-----------------------------------------------------------------------------
+
+//-----------------------------------------------------------------------------
+// Simple nodetype for identifying the login page.
```
```suggestion
// Simple nodetype holding information about the lock.
```

sdumitriu commented on modules/locking/src/main/resources/SLING-INF/content/Extensions.json:
Duplicate, remove this file?

sdumitriu commented on modules/locking/src/main/resources/SLING-INF/content/Extensions/SubjectActions/subjectLockAction.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "jcr:primaryType": "cards:Extension",
+    "cards:extensionPointId": "cards/coreUI/actions/subject",
```
This isn't the expected ID.

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockedRestrictionPattern.java:
```diff
@@ -0,0 +1,132 @@
+import org.apache.jackrabbit.oak.spi.security.authorization.restriction.RestrictionPattern;
+
+/**
+ * A restriction that makes a permissions entry only be valid on nodes created by the accessing user.
```
Update.

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockServlet.java:
```diff
@@ -0,0 +1,324 @@
+@SlingServletResourceTypes(
+    resourceTypes = { "cards/Subject" },
+    methods = { "LOCK", "UNLOCK" })
+public class LockServlet extends SlingSafeMethodsServlet
```
These aren't safe methods, it should extend `SlingAllMethodsServlet`.

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockServlet.java:
```diff
@@ -0,0 +1,324 @@
+    public void handleRequest(final SlingHttpServletRequest request, final SlingHttpServletResponse response,
+        final boolean isLockRequest)
+        throws IOException, IllegalArgumentException
+    {
```
The actual functionality should be in a separate class, a service usable from other code, and the servlet should just forward calls to the service:

- add interface `io.uhndata.cards.locking.api.LockManager` with `isLocked`, `canLock`, `tryLock`, `forceLock`, `canUnlock`, `unlock` methods
- add implementation `io.uhndata.cards.locking.internal.LockManagerImpl`
- `tryLock` checks if any preconditions are not satisifed; this checks if any forms are still incomplete, and if yes, fail the operation and return a warning message
- `forceLock` just locks anyway
- add error classes `io.uhndata.cards.locking.api.LockWarning` and `LockError`, with a simple message property
- add interface `io.uhndata.cards.locking.spi.LockPrecondion` with a methods `Boolean canLock(Node n) throws LockWarning, LockError` and `canUnlock`
- add `io.uhndata.cards.locking.internal.IncompleteFormsPrecondition` which checks if n is a subject and has incomplete forms
- future, add `io.uhndata.cards.locking.internal.LockRightPrecondition` which checks if the current user has the right to lock
- here add a dependency on the service and just call the service

sdumitriu commented on modules/locking/src/main/java/io/uhndata/cards/locking/LockedRestrictionPattern.java:
```diff
@@ -0,0 +1,132 @@
+        if (isForm(node)) {
+            String subject = node.getProperty("subject").getValue(Type.REFERENCE);
+            try {
+                return isLocked(this.session.getNodeByIdentifier(subject));
```
Does this handle these cases, in LFS mode:
- create Patient > Tumor > Tumor Region
- lock the patient
- create a form for the tumor region
- create a new Tumor Region under the existing Tumor
- create a new Tumor Region and a new Tumor parent for the existing Patient

acrowthe replied:
Tested, all of those creations are blocked successfully due to the locked parents

sdumitriu commented:
I think this will need adjusting, but not right now: creating a new form from the dashboard should not allow selecting a locked subject. I think this would be better handled by checking rights themselves, instead of customizing even more the subject selector component.

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
Refactor lock handling into LockManager rather than built into LockServlet

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
Remove extraneous lock action extension definition

acrowthe committed:
CARDS-2514: New "locking" module, with basic locking feature
Update locking module version

acrowthe commented:
Replaced by #1881
